### PR TITLE
Add automatic moderation

### PR DESCRIPTION
**Pull request by fatlotus**
Currently, moderations are approved by a human moderator (myself), who is hoping to graduate this quarter. The Auto-Moderator is a toggleable listing and inquiry approver that simulates a human moderator with some dumb blacklisting heuristics. By running it every hour, we can hopefully reduce the volume of listings to moderate.

### CODE REVIEW

File: caravel/controllers/moderation.py [hunk 1]
```diff
@@ -4,11 +4,27 @@
...
 from google.appengine.api import users
 
 
+class ModerationConfig(ndb.Model):
+    """
+    Defines the static configuration of the automoderator.
+    """
+
+    @classmethod
+    def get(klass):
```
Comment (fatlotus) [inline caravel/controllers/moderation.py]: In this case, there's only ever supposed to be one ModerationConfig — so this step allows us to get that (singleton) entity.

File: caravel/controllers/moderation.py [hunk 2]
```diff
@@ -50,6 +68,15 @@ def moderation_view():
         key.delete()
         return ""
 
+    if request.form.get("automod"):
+        if config.enabled:
```
Comment (fatlotus) [inline caravel/controllers/moderation.py]: Moderation.enabled is the property, `Moderation.get().enabled` is the value of that property on the singleton instance. (`Moderation.get()` is a database call to retrieve the configuration, so we don't want to do it multiple times.)

File: caravel/controllers/moderation.py [hunk 3]
```diff
@@ -58,4 +85,42 @@ def moderation_view():
...
+    for to_moderate in pending:
+        if to_moderate.age < datetime.timedelta(minutes=config.min_delay):
+            continue
+
+        values = dict(to_moderate.to_dict())
+        values["principal"] = values["principal"].email
+        allowed = True
+
+        for field in values.values():
+            for item in config.blacklist:
```
Comment (fatlotus) [inline caravel/controllers/moderation.py]: Repeated database properties are the empty list by default. I can see where that's not really clear, though. Ideas for better commenting?

**Comment by fatlotus** [comment]
Bump!

File: caravel/controllers/moderation.py (hunk 3 above)
Comment (georgeteo) [inline caravel/controllers/moderation.py]: Maybe put a note here: https://github.com/uchicago-sg/caravel/pull/239/files#diff-ad46fd1aeff1d04636c82ea26be28274R24

PS: is there a better way of referencing code numbers in comments?

**Comment by georgeteo** [comment]
Looks good. Merge away once you've responded/fixed the reply on the last comment.

File: caravel/controllers/moderation.py (hunk 3 above)
Comment (fatlotus) [inline caravel/controllers/moderation.py]: Added a note about the default value of `blacklist`.

**Pull request merged**